plugins/inputs/pf/pf.go

```diff
@@ -0,0 +1,207 @@
+	"strconv"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const measurement = "pf"
+const pfctlCommand = "pfctl"
+
+type PF struct {
+	UseSudo    bool
+	StateTable StateTable
+	infoFunc   func() (string, error)
+}
+
+func (pf *PF) Description() string {
+	return "Gather counters from PF"
+}
+
+func (pf *PF) SampleConfig() string {
+	return `
+  ## PF require root access on most systems.
+  ## Setting 'use_sudo' to true will make use of sudo to run pfctl.
+  ## Users must configure sudo to allow telegraf user to run pfctl with no password.
+  ## pfctl can be restricted to only list command "pfctl -s info".
+  use_sudo = false
+`
+}
+
+// Gather is the entrypoint for the plugin.
+func (pf *PF) Gather(acc telegraf.Accumulator) error {
+	o, err := pf.infoFunc()
+	if err != nil {
+		acc.AddError(err)
+		return nil
+	}
+
+	if perr := pf.parsePfctlOutput(o, acc); perr != nil {
+		acc.AddError(perr)
+	}
+	return nil
+}
+
+var errParseHeader = fmt.Errorf("Cannot find header in %s output", pfctlCommand)
+
+func errMissingData(tag string) error {
+	return fmt.Errorf("struct data for tag \"%s\" not found in %s output", tag, pfctlCommand)
+}
+
+var stateTableHeaderRE = regexp.MustCompile("^State Table")
+var anyTableHeaderRE = regexp.MustCompile("^[A-Z]")
+
+func (pf *PF) parsePfctlOutput(pfoutput string, acc telegraf.Accumulator) error {
+	lines := strings.Split(pfoutput, "\n")
+	stateTableFound := false
+	for i, line := range lines {
+		if stateTableHeaderRE.MatchString(line) {
+			endline := len(lines)
+			for j, el := range lines[i+1:] {
+				if anyTableHeaderRE.MatchString(el) {
+					endline = j + i + 1
+					break
+				}
+			}
+			if perr := pf.parseStateTable(lines[i+1:endline], acc); perr != nil {
+				return perr
+			}
+			stateTableFound = true
+		}
+	}
+	if !stateTableFound {
+		return errParseHeader
+	}
+	return nil
+}
+
+var stateTableRE = regexp.MustCompile(`^  (.*?)\s+(\d+)`)
+
+func (pf *PF) parseStateTable(lines []string, acc telegraf.Accumulator) error {
+	st := StateTable{}
+	tags, err := st.getTags()
+	if err != nil {
+		return fmt.Errorf("Can't retrieve struct tags: %v", err)
+	}
+	fMap := make(map[string]bool)
+	for i := 0; i < len(tags); i++ {
+		fMap[tags[i]] = false
+	}
+
+	for _, v := range lines {
+		entries := stateTableRE.FindStringSubmatch(v)
+		if entries != nil {
+			fs, err := st.setByTag(entries[1], entries[2])
+			if err != nil {
+				return errors.New("can't set statetable field from tag")
+			}
+			if fs {
+				fMap[entries[1]] = true
+			}
+		}
+	}
+
+	for k, v := range fMap {
+		if !v {
+			return errMissingData(k)
+		}
+	}
+
+	fields := make(map[string]interface{})
+	fields["entries"] = st.CurrentEntries
+	fields["searches"] = st.Searches
+	fields["inserts"] = st.Inserts
+	fields["removals"] = st.Removals
+	acc.AddFields(measurement, fields, make(map[string]string))
+	return nil
+}
+
+func (pf *PF) callPfctl() (string, error) {
+	c, err := pf.buildPfctlCmd()
```

Comment:
This should only be ran once on startup.

Reply:
will fix